samza-azure/src/main/java/org/apache/samza/AsyncSystemProducer.java

```diff
@@ -0,0 +1,169 @@
+  private final HashMap<String, SamzaHistogram> sendCallbackLatency = new HashMap<>();
+  private final HashMap<String, Counter> sendErrors = new HashMap<>();
+
+  /**
+   * Instantiates a new Async system producer.
+   *
+   * @param systemName the system name
+   * @param config the config
+   * @param registry the registry
+   */
+  public AsyncSystemProducer(String systemName, Config config, MetricsRegistry registry) {
+    StreamConfig sconfig = new StreamConfig(config);
+    streamIds = config.getList(String.format(CONFIG_STREAM_LIST, systemName));
+    physicalToStreamIds =
+        streamIds.stream().collect(Collectors.toMap(sconfig::getPhysicalName, Function.identity()));
+    this.registry = registry;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public void send(String source, OutgoingMessageEnvelope envelope) {
+    checkCallbackThrowable("Received exception on message send");
+    CompletableFuture<Void> sendResult = sendAsync(source, envelope);
+    pendingFutures.add(sendResult);
+
+    String streamName = envelope.getSystemStream().getStream();
+    String streamId = physicalToStreamIds.getOrDefault(streamName, streamName);
+    long afterSendTimeMs = System.currentTimeMillis();
+
+    // Auto update the metrics and possible throwable when futures are complete.
+    sendResult.handle((aVoid, throwable) -> {
+        long callbackLatencyMs = System.currentTimeMillis() - afterSendTimeMs;
+        sendCallbackLatency.get(streamId).update(callbackLatencyMs);
+        aggSendCallbackLatency.update(callbackLatencyMs);
+        if (throwable != null) {
+          sendErrors.get(streamId).inc();
+          aggSendErrors.inc();
+          LOG.error("Send message to event hub: {} failed with exception: ", streamId, throwable);
+          sendExceptionOnCallback.compareAndSet(null, throwable);
+        }
+        return aVoid;
+      });
+  }
+
+  public void start() {
+    streamIds.forEach(streamId -> {
+        sendCallbackLatency.put(streamId, new SamzaHistogram(registry, streamId, SEND_CALLBACK_LATENCY));
+        sendErrors.put(streamId, registry.newCounter(streamId, SEND_ERRORS));
+      });
+
+    aggSendCallbackLatency = new SamzaHistogram(registry, AGGREGATE, SEND_CALLBACK_LATENCY);
+    aggSendErrors = registry.newCounter(AGGREGATE, SEND_ERRORS);
+  }
+
+  /**
+   * Sends a specified message envelope from a specified Samza source.
+   * @param source String representing the source of the message.
+   * @param envelope Aggregate object representing the serialized message to send from the source.
+   * @return the completable future for the async send call
+   */
+  public abstract CompletableFuture<Void> sendAsync(String source, OutgoingMessageEnvelope envelope);
+
+  /**
+   * Check callback throwable.
+   *
+   * @param msg the msg
+   */
+  protected void checkCallbackThrowable(String msg) {
+    // Check for send errors from EventHub side
+    Throwable sendThrowable = sendExceptionOnCallback.get();
```

Comment:
When there is an exception on callback, Only way to clear it is to recreate the system producer. Because it is possible that the connection is broken.